src/referral_maintenance.rs

```diff
@@ -27,20 +41,159 @@ pub async fn maintenaince_tasks(db: Arc<ReferralStore>, dune_data_folder: String
-            Ok(_) => {}
-            Err(err) => tracing::debug!("Error during maintenaince_task for referral: {:?}", err),
+async fn download_referral_from_ipfs_and_store_in_referral_store(
+    db: Arc<ReferralStore>,
+    hash: H256,
+) -> Result<()> {
+    match get_cid_from_app_data(hash) {
+        Ok(cid) => {
+            tracing::debug!("cid for hash {:?} is {:?}", hash, cid);
+            match get_ipfs_file_and_read_referrer(cid.clone()).await {
+                Ok(referrer) => {
+                    tracing::debug!("Adding the referrer {:?} for the hash {:?}", referrer, hash);
+                    let mut guard = match db.0.lock() {
+                        Ok(guard) => guard,
+                        Err(poisoned) => poisoned.into_inner(),
+                    };
+                    guard.app_data.insert(hash, Referral::Address(referrer));
+                }
+                Err(err) => {
+                    tracing::debug!(
+                        "Could not find referrer in cid {:?}, due to the error {:?}",
+                        cid,
+                        err
+                    );
+                    let mut guard = match db.0.lock() {
+                        Ok(guard) => guard,
+                        Err(poisoned) => poisoned.into_inner(),
+                    };
+                    guard.app_data.entry(hash).and_modify(|referral_entry| {
+                        *referral_entry = match referral_entry.clone() {
+                            Referral::TryToFetchXTimes(x) => {
+                                if x > 1u64 {
+                                    Referral::TryToFetchXTimes(x - 1)
+                                } else {
+                                    Referral::Address(None)
+                                }
+                            }
+                            _ => Referral::Address(None),
+                        }
+                    });
+                }
+            }
         }
-        tokio::time::sleep(MAINTENANCE_INTERVAL).await;
+        Err(err) => {
+            tracing::debug!("For the app_data hash {:?}, there could not be found a unique referrer due to {:?}", hash, err);
+        }
+    }
+    Ok(())
+}
+
+async fn get_ipfs_file_and_read_referrer(cid: String) -> Result<Option<H160>> {
+    // Front-end is uploading the ipfs to pinta, hence we could also get the
+    // data from "https://gateway.pinata.cloud/ipfs/{:}", though they are rate limiting us.
+    let url = format!("https://ipfs.io/ipfs/{:}", cid);
+    let client = reqwest::Client::builder()
+        .timeout(Duration::from_secs(1))
+        .build()?;
+    let body = client.get(url.clone()).send().await?.text().await?;
+    let json: AppData = serde_json::from_str(&body)?;
+    if let Some(metadata) = json.metadata {
+        if let Some(referrer) = metadata.referrer {
+            return Ok(Some(referrer.address));
+        }
+    }
+    Ok(None)
```

Comment:
I'm trying to become a rust guru and this kind of nested block strikes me as something that could probably be written in a different way. Although I imagine the combined `Result<Option<_>>` is whats making it difficult to approach here.

Reply:
The best I ccan think of is:
```rust
Ok(json.metadata.and_then(|metadata| Some(metadata.referrer?.address)))
```